src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -644,6 +644,82 @@ protected override void ApplyAnalyzerReferenceRemoved(ProjectId projectId, Analy
             }
         }
 
+        protected override void ApplyProjectChanges(ProjectChanges projectChanges)
+        {
+            base.ApplyProjectChanges(projectChanges);
+
+            var changedMappedDocuments = projectChanges.GetChangedDocuments().Where(id => IsMappedDocument(id));
+            if (changedMappedDocuments.Any())
+            {
+                ApplyChangesForMappedDocuments(projectChanges, changedMappedDocuments);
+            }
+
+            bool IsMappedDocument(DocumentId documentId)
+            {
+                return this.CurrentSolution.GetDocument(documentId)?.Services?.GetService<ISpanMappingService>() != null;
+            }
+        }
+
+        private void ApplyChangesForMappedDocuments(ProjectChanges projectChanges, IEnumerable<DocumentId> changedMappedDocuments)
+        {
+            // Map all the original text changes to new text changes in the mapped files.
+            var _ = ArrayBuilder<(string fileName, TextChange textChange)>.GetInstance(out var mappedChangesBuilder);
+            foreach (var changedDocumentId in changedMappedDocuments)
+            {
+                var oldDoc = projectChanges.OldProject.GetDocument(changedDocumentId)!;
+                var newDoc = projectChanges.NewProject.GetDocument(changedDocumentId)!;
+
+                GetMappedTextChanges(newDoc, oldDoc, mappedChangesBuilder);
+            }
+
+            var mappedChanges = mappedChangesBuilder.ToImmutable();
+            if (mappedChanges.Any())
+            {
+                // Group the mapped text changes by file, then apply all mapped text changes to each file.
+                foreach (var changesForFile in mappedChanges.GroupBy(change => change.fileName))
+                {
+                    using var invisibleEditor = new InvisibleEditor(ServiceProvider.GlobalProvider, changesForFile.Key, GetHierarchy(projectChanges.ProjectId), needsSave: true, needsUndoDisabled: false);
+                    TextEditApplication.UpdateText(changesForFile.Select(change => change.textChange), invisibleEditor.TextBuffer, EditOptions.None);
+                }
+            }
+
+            void GetMappedTextChanges(CodeAnalysis.Document newDocument, CodeAnalysis.Document oldDocument, ArrayBuilder<(string, TextChange)> builder)
+            {
+                var mappingService = newDocument.Services.GetService<ISpanMappingService>();
+                (var mappedSpanResults, var textChanges) = _threadingContext.JoinableTaskFactory.Run<(ImmutableArray<MappedSpanResult>, ImmutableArray<TextChange>)>(async () =>
```

Comment:
How many files are likely to be in `changedMappedDocuments` above? Is it better to gather all of this data in one go, to not thrash around the thread pool as much?

Reply:
Good point - it could be many depending on how many files the identifier is present in.  I'll make this retrieve all the mapping results in one jtf.Run